daemon/containerd/image_list.go

```diff
@@ -77,29 +78,48 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 			continue
 		}
 
-		platforms, err := images.Platforms(ctx, contentStore, img.Target)
-		if err != nil {
-			return nil, err
-		}
-
-		for _, platform := range platforms {
-			image, chainIDs, err := i.singlePlatformImage(ctx, contentStore, img, platform)
-			if err != nil {
-				return nil, err
-			}
-			if image == nil {
-				continue
+		err := images.Walk(ctx, images.HandlerFunc(func(ctx context.Context, desc v1.Descriptor) ([]v1.Descriptor, error) {
+			if images.IsIndexType(desc.MediaType) {
+				return images.Children(ctx, contentStore, desc)
 			}
 
-			summaries = append(summaries, image)
+			if images.IsManifestType(desc.MediaType) {
+
+				// Ignore buildkit attestation manifests
+				// https://github.com/moby/buildkit/blob/master/docs/attestations/attestation-storage.md
+				if _, has := desc.Annotations[attestation.DockerAnnotationReferenceType]; has {
+					return nil, nil
+				}
+
+				if desc.Platform == nil {
+					return nil, nil
+				}
+
+				image, chainIDs, err := i.singlePlatformImage(ctx, contentStore, img, *desc.Platform)
+				if err != nil {
+					return nil, err
+				}
+				if image == nil {
```

Comment:
Hm... we should look at this, it's a weird interface, and feels like "unexpected" to not have an error, but no content either;
https://github.com/moby/moby/blob/a1f6a705225eedcb55fe6dac7dec0b9870330d78/daemon/containerd/image_list.go#L119-L123

Wondering if we can change this to return a `NotFound()` or something else to indicate we cannot use the result. (at least a GoDoc somewhere explaining _that_ it may happen, and "when" would be good). If I'm not mistaken the `nil` can only happen in that branch, in all other cases we will have something returned.

Reply:
I hoisted the check up so we don't have to return the error now.